### ISSUE TITLE
[Power Up] Quantum Portfolio Analysis

### ISSUE THREAD

**Issue by calumholker** [XanaduAI/QHack2021#20]
### Team Name: 

QLords

### Project Description: 

As a team of undergraduates that are relatively new to quantum machine learning we are going to explore different methods of applying QML to the mean variance portfolio optimisation problem for N assets. The main focus will be to implement a qGAN. There are papers on how classical GANs can be applied to portfolio optimisation (https://arxiv.org/pdf/1909.10578.pdf) and papers on the implementation of qGANs. The main paper we will be using to implement this is - (https://www.nature.com/articles/s41534-019-0223-2.pdf).

The method is as follows. Firstly we extract the data required from the internet and pass it into computable sequences. We have a quantum generator that takes a time series as input, and outputs the next period for the time series. We train this using a classical CNN discriminator. This takes in a sequence and returns if it is generated or real. We then can use the quantum generator and classical CNN to train each other. Once they have been trained the generator will be able to produce the next period. This can then be analysed to maximise gain and minimise risk.

Due to the limits of current quantum computation, we will be exploring the method we would use given larger computers. This may mean that the results are not significant when compared to the use of a classical GAN, however we think that this is a more informative project for how real data can be used. We are going to do this by passing large datasets through, as you would if you were using a large classical NN for time series analysis. This means that although we may not have the time to train the model enough (or with enough datasets) such that the results are comparable, we can see how the results begin to improve, and theorise the extrapolation given more time and computational power.

Time permitting, besides qGANs we are also going to look at two other methods of portfolio analysis. The first will be QAOA to find the optimal solution, creating a hamiltonian for the problem.

We are further going to look into using quantum inspired Tensor Networks - this involves mapping out the optimisation problem using the historical data into a QUBO and then transform it into an MPS, then run DMRG into it. i.e Make an MPS and Hamiltonian for the problem, find the minimal value of the Hamiltonian with DMRG that is simultaneously solving the MPS from two ends keeping one state constant, one iteration at a time. This method can be fed into the qGAN and trained alongside the qGAN.

### Source code: 

Initial planning and some rough implementations:
https://github.com/calumholker/quantum-portfolio-optimisation

### Resource Estimate: 

We will require the maximum number of qubits in order to pass in as large datasets as possible. In order to obtain the best results we can, the maximum amount of data will need to be processed in the minimum amount of time. We are therefore using the FLOQ simulator, as this takes a fraction of the time to compute our generator circuit. Once testing is complete we will train our qGAN on an Amazon Braket machine. As for our project the longer we can spend computing the better the results will be, we will use all of our available budget in training the model with the large datasets.

In terms of the datasets, we have extracted 22350 sequences of length 40 from historical data for an array of different stocks. Each of these will be split into a 32 length and 8 length, the 32 being used to generate and the 8 being used to train the discriminator. Each of these computations takes on the order of half an hour on a local simulator, compared to 45 seconds on FLOQ. We will obviously not be able to run all 22350 on the machine, as several training loops are required for each. Instead we are going to take on the order of 100 sequences for one stock and show how this model can be used on this stock. If we complete this we will pass more data to see if it improves.

**Comment by co9olguy** [comment]
Thanks for the draft submission @calumholker! :smiley: 

There is still some time left before the deadline. In order to give your project the best chance at success, I would recommend fleshing out the descriptions a bit more. In particular, the Resource Estimate will be crucial for determining which entries will receive the Power Up. The more you can give details about how those credits will be applied specifically, the better your submission will look.

**Comment by co9olguy** [comment]
Thanks for your Power Up Submission @calumholker !

To help us keep track of final submissions, we will be closing all of the [Power Up] issues. We ask you to open a new issue for your final submission. Please use [this pre-formatted [Entry] Issue template](https://github.com/XanaduAI/QHack/issues/new?assignees=&labels=submission&template=open-hackathon-final-submission.md&title=%5BENTRY%5D+Your+Project+Title). Note that for the final submission, the Resource Estimate requirement is replaced by a Presentation item.
